addons/sale_margin/models/sale_order_line.py

```diff
@@ -20,20 +20,21 @@ class SaleOrderLine(models.Model):
     @api.depends('product_id', 'company_id', 'currency_id', 'product_uom')
     def _compute_purchase_price(self):
         for line in self:
```

Comment:
```suggestion
        for line in self.filtered(
            lambda sol: not (sol.purchase_price and sol.state in ['sale', 'cancel'])
        ):
```
Filtering them out beforehand is usually more efficient ORM-wise, plus you don't have diff every following line.